12326 export missing models org exporter

javitonino opened the pull request:
Addresses some of the issues in #12326:
- Groups export
- Notifications export
- Importing a viewer users with visualizations
- Dataservices usage (from redis)
- Rakes to make migration easier (also see https://github.com/CartoDB/cartodb-central/pull/1799). Documented in https://github.com/CartoDB/doc-internal/pull/17

**Acceptance note**: You have to migrate between clouds, or ensure that you delete redis keys if you do it in a single machine (export + delete + import).

Also fixes #12364

juanignaciosl commented on app/models/carto/group.rb:
```diff
@@ -24,9 +24,7 @@ class Group < ActiveRecord::Base
-    private_class_method :new
-
-    validates :name, :database_role, :organization_id, presence: true
+    validates :name, :database_role, :organization, presence: true
```
I'd like to keep `private_class_method :new`. As creating a group involves logic at DB level, keeping construction under control with factory methods is a must. Maybe moving `build_group_from_hash` here is enough, as long as it actually works.

PS: have you checked that importing groups work? Share a dataset, RW mode, and check that group users can edit.

javitonino replied:
There should be no problem moving the function into the model. I considered it when I saw the private constructor, but I'm not a fan of it. The reason is that it is pretty much the same thing, since this functions just calls the constructor directly, with no transformation done to the input parameters. The good thing is that if you try to use `.new` you will get an error, so you are forced to take a look at the class to look at the `create_*` methods. So maybe we can migrate it and add a comment saying to be careful with this one, as it won't do the DB operations.

juanignaciosl replied:
Yep, that's it. Even though the method is merely an alias, it makes the issue explicit (although it needs a brief explanation).

juanignaciosl replied:
You forgot bringing `private_class_method :new` back.

javitonino replied:
I'm dangerously distracted :/

juanignaciosl commented on app/services/carto/organization_metadata_export_service.rb:
```diff
@@ -38,6 +38,8 @@ def build_organization_from_hash_export(exported_hash)
     end
 
     def save_imported_organization(organization)
+      organization.groups.each { |g| g.users_group.clear }
```
Why is this needed? Maybe a comment would be nice, the need for clearance before saving is not obvious.

javitonino replied:
The comment is in the caller: https://github.com/CartoDB/cartodb/pull/12380/files/5e52fe903180ad0eae2b4490f55376b2aa8a8806#diff-864e4f8e01b274ac9e923f3adaa33ae5R199. Probably should be another comment here.

About the logic: group users cannot be saved before users are themselves imported. So the import order is organization -> users -> groups. This function is intended to be called in two ways:
- Directly, to only import organization metadata. In that case, I try to keep the groups (without users)
- From `import_organization_and_users_from_directory`, which imports everything. In that case, we do the proper saving order as above.

Maybe we can remove the support for the first call (make this method private), as I don't think it will be used in the production code (right now it's only used in tests), that way, we could remove the groups clearing from this function completely (and do it in the caller). This implies changes to the test, but such is life.

juanignaciosl replied:
If that's the case, please do. It seems a lot of complication to enable testing this. As long as we have tests covering the full case, there's no need for tests around this method.

juanignaciosl commented on app/services/carto/organization_metadata_export_service.rb:
```diff
@@ -95,6 +133,34 @@ def export_asset(asset)
+        display_name: group.display_name,
+        database_role: group.database_role,
+        auth_token: group.auth_token,
+        users: group.users.map(&:id)
```
Maybe `user_ids` is more accurate?

juanignaciosl commented:
This seems to fail for viewers with visualizations:

```
{:user=>"Viewer users can't store visualizations"}

Tasks: TOP => cartodb:user_migrator:import:organization
(See full trace by running task with --trace)
```

I've tested moving "j-ded-06-deleteme" from ded06 to ded04, I'll keep code there so you can take a look if you want, @javitonino .

javitonino commented:
Fixed. I cleaned the failed import (`org.destroy_cascade(delete_in_central: false)`, plus cleaning redis keys) and deployed to ded04 for you, so you can continue testing.

juanignaciosl commented:
There's another minor issue: importing notifications duplicates the warning. STR:

1. Create a notification for everybody, and read it with a user.
2. Export and import.
3. Users that didn't read it have the notification duplicated, and those who did, once.

mayoral commented:
Created an organization in ded02, and moved it without any issues to ded06. All the things (DNS setup, export - import, organization move) worked like a charm!

Acceptance status: 👌
